lib/ansible/modules/windows/win_chocolatey.ps1

```diff
@@ -432,10 +446,53 @@ Function Choco-Uninstall
     $result.changed = $true
 }
 
+Function Choco-ConfigureProxy
+{
+    [CmdletBinding()]
+
+    param(
+        [Parameter(Mandatory=$false, Position=1)]
+        [string]$proxy_url,
+        [Parameter(Mandatory=$false, Position=2)]
+        [string]$proxy_username,
+        [Parameter(Mandatory=$false, Position=3)]
+        [string]$proxy_password
+    )
+    $hash = @{
+        proxy = $proxy_url
+        proxyUser = $proxy_username
+        proxyPassword = $proxy_password
+    }
+    foreach ($h in $hash.GetEnumerator()) {
+        if ($($h.Value))
+        {
+          $cmd = "$executable config set $($h.Name) $($h.Value)"
+        }
+        else
+        {
+          $cmd = "$executable config unset $($h.Name)"
+        }
+        $results = Invoke-Expression $cmd
+        if ($LastExitCode -ne 0)
+        {
+            Set-Attr $result "choco_error_cmd" $cmd
+            Set-Attr $result "choco_error_log" "$results"
+
+            Throw "Error setting $($h.Name) with $($h.Value)"
+        }
+        If ("$results" -notmatch "Nothing to change. Config already set.")
+        {
+            $result.changed = $true
+        }
+    }
+}
+
 Try
 {
     Chocolatey-Install-Upgrade
 
+    Choco-ConfigureProxy
```

Comment:
While I don't think you should use this as it changes the global config, it currently doesn't work as it needs you to set `-proxy_url`, `-proxy_username`, `-proxy_password` as it is expecting those params. I don't think the `choco config set` and `choco config unset` should be used as it is a global setting and users may be unaware that setting this can affect it when they run it locally

Instead I think you should use the explicit proxy settings at runtime as shown here https://chocolatey.org/docs/proxy-settings-for-chocolatey#installing-chocolatey-behind-a-proxy-server. It does have a dependency on a relatively new version of chocolatey but I think the benefits outweigh the negatives here and we can just simply say in the documentation these parameters require version `0.10.4` or newer.

If you feel really adventurous you can try and parse the version and fail if it is older and the proxy params are set.

Reply:
My concerns are with not using standard parameter names and values for the proxy. I would much rather see **proxy** and **no_proxy** options and the same behavior as other modules (using the system-wide proxy settings, if set).

Reply:
Agree with that @dagwieers but I think it is too late to work in the `use_proxy` option, currently these changes work with the currently set system proxy so at least we have the default action there. It shouldn't be too hard to work in the `use_proxy: no` option to disable the system proxy but prefer to leave that for another PR.

Reply:
Great thing if proxy support can make its way into 2.4 , Thanks @jborean93

Not modifying global chocolatey configuration looks neat too. Support for `--proxy-bypass-list` and `--proxy-bypass-on-local` parameters should be added too.

Reading the docs at https://chocolatey.org/docs/proxy-settings-for-chocolatey#installing-chocolatey-behind-a-proxy-server I realize that with chololatey  0.10.4+ versions, proxy should also work through the use of the following env variables : `http_proxy`, `https_proxy`, `no_proxy` making it easier to work with proxies through ansible than it was before.

Also one question comes to mind, what happens if a proxy is globally set on chocolatey install but we don't want to use one through ansible. (weird improbable case I admit)

Reply:
I already mentioned this, that's what **no_proxy** is for. Look at the **win_get_url** module. The **proxy_url** parameter should be renamed to **proxy** as that's the standard name used by other modules.